File: 19th/코딩테스트 심화반/week6/1167번_트리의 지름.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<pair<int, int>> adj[100001]; // 트리를 인접리스트로 저장
bool visited[100001]; // 방문 여부

int max_dist_n = 1; // 트리의 지름이 되는 끝 노드, 초기값은 임의의 노드로 설정
int max_dist = 0; // 트리의 지름

void dfs(int n, int dist) {
	for (int i = 0; i < adj[n].size(); i++) {	
		if (!visited[adj[n][i].first]) { // 방문하지 않았을 경우
			if (max_dist < dist + adj[n][i].second) { // 현재까지의 거리가 더 클 경우
				max_dist_n = adj[n][i].first; 
				max_dist = dist + adj[n][i].second;
			}

			visited[adj[n][i].first] = true;
			dfs(adj[n][i].first, dist + adj[n][i].second);
		}		
	}
}

int main() {
	int V;
	cin >> V;

	int n, nn, d;
	for (int i = 0; i < V; i++) {
		cin >> n;
		while (1) {
			cin >> nn;
			if (nn == -1) break;
			cin >> d;
			
			adj[n].push_back({nn, d});
		}
	}

	visited[max_dist_n] = true;
	dfs(max_dist_n, 0);

	fill(visited, visited + V + 1, false); // visited 배열 초기화
	visited[max_dist_n] = true;
	dfs(max_dist_n, 0);

	cout << max_dist;
}

File: 19th/코딩테스트 심화반/week6/1167번_트리의 지름.py
import sys
sys.setrecursionlimit(200000)

v = int(input())

# 인접 리스트 생성
adj = [[] for _ in range(v + 1)]
for _ in range(v):
    readline = list(map(int, input().split()))
    for i in range(len(readline) // 2 - 1):
        adj[readline[0]].append((readline[2 * i + 1], readline[2 * i + 2]))
        adj[readline[2 * i + 1]].append((readline[0], readline[2 * i + 2]))

idx, max_dist = 1, 0
visited = [0] * (v + 1)
visited[1] = 1

def dfs(n, dist):
    global idx, max_dist
    global visited

    # 현재 노드 n에서 갈 수 있는 모든 인접 노드 탐색
    for i, j in adj[n]:  # (i: 다음 노드, j: 거리)
        if not visited[i]:
            visited[i] = 1
            sum_dist = dist + j
            if sum_dist > max_dist:  # 더 멀리 도달하면 갱신
                max_dist = sum_dist
                idx = i
            dfs(i, dist + j)

# 1단계: 임의의 노드(1)에서 가장 먼 노드 찾기
dfs(1, 0)

# 2단계: 그 노드(idx)에서 가장 먼 노드를 찾아 거리 측정 → 트리의 지름
visited = [0] * (v + 1)
visited[idx] = 1
dfs(idx, 0)

print(max_dist)